cpp/src/toppra/parametrizer/const_accel.cpp

```diff
@@ -61,6 +61,13 @@ Vectors ConstAccel::eval_impl(const Vector& times, int order) const {
   Vector ss, vs, us;
   TOPPRA_LOG_DEBUG("eval_impl. order=" << order);
   bool ret = evalParams(times, ss, vs, us);
+  auto path_interval = m_path->pathInterval();
+  // Clamp all elements of ss to within the bounds of the path
+  for (std::size_t i = 0; i < ss.size(); i++) {
+    if (ss[i] > path_interval[1]) {
+      ss[i] = path_interval[1];
+    }
+  }
```

Comment:
I am fine with clamping numerical errors. The way it is done here is non-friendly for debugging because it doesn't fail or even warn when the elements are far above the limit.

Reply:
Do you think clamping within some epsilon would work here? I'm not at a PC right now but I seem to remember there exists a TOPPRA_EPS which could be used.

Reply:
You can try with this epsilon. It may not be sufficient in your case since the path length is very big.

Reply:
The large path is just a canned example (see comment below).

Another option might be to throw early if any element in times is outside the constaccel bounds, similar to the check in PiecewisePoly. Then the clamping here would only be done if a numerical error has moved a value outside of the bounds of the source path.

Yet another option would be to require a source path that supports extrapolation, but this seems like it would be an unexpected.

Reply:
I am fine with clamping. You can do something like:
```
if ((ss.array() < path_interval[0] * (1 - EPS)).any())
  throw std::runtime_error("a meaningful message");
if ((ss.array() > path_interval[1] * (1 + EPS)).any())
  throw std::runtime_error("a meaningful message");
ss = ss.cwiseMax(path_interval[0]).cwiseMin(path_interval|1]);
```
Usually, the best is to mix relative and absolute test. The above test may not work find when 0 (or close to 0) is one bound of the path interval. If the path interval bound is inferior to EPS, then you may keep EPS.

Reply:
Thanks for the reply - I'm not at a PC right now but I'll take a look in a few hours.

Reply:
I think what @jmirabel means is to use `EPS_REL` and `EPS_ABS` in the check, i.e., 
```
if ((ss.array() < (path_interval[0] * (1 - EPS_REL) - EPS_ABS)).any())
  throw std::runtime_error("a meaningful message");
```

`EPS_ABS` can be given some meaningful values such as 0.1ms (1-e4).

Reply:
I don't quite understand the advantage of doing it this way. Surely, if a user asks for a number inside `ConstAccel::pathInterval()`, this should always work, and outside should always fail (with a meaninful message). It seems that EPS doesn't come into it at all.

Reply:
The sort of scenarios that this code can help catch is numerical errors. Say for example, suppose you generate several time instances that should span [0, 3.1423231], I would expect that the last time instance can be slight _off_.

It's therefore reasonable to clamp if the numerical checks with eps pass. Otherwise, the code should throw exceptions on widely off values (-1 in a [0, 1.0] path).

Reply:
I feel like it should be up to the caller to make sure their values are exactly within the reported bounds. At least this is the approach taken by the `PiecewisePolyPath` object:

piecewise_poly_path.cpp:148
```cpp
if (pos > m_breakpoints[m_breakpoints.size()-1] || pos < m_breakpoints[0]) {
  std::ostringstream oss;
  oss << "Position " << pos << " is outside of range [ " << m_breakpoints[0]
    << ", " << m_breakpoints[m_breakpoints.size()-1] << ']';
  throw std::runtime_error(oss.str());
}
```
I'm not adverse to the EPS approach, I just don't think it's necessary. Your call though.

Reply:
> I feel like it should be up to the caller to make sure their values are exactly within the reported bounds.

Isn't it what is implemented in the current HEAD. You get an exception if you are not exactly within bounds. And this is what your PR tries to address...

relative + absolute check make sense.

Reply:
That's right, the current HEAD fixes my issue and the test I have written asserts this. I evidently haven't done a good job at explaining what this PR is trying to address, so I'll rename my PR for clarity. I'm not trying to protect the caller from themselves, rather I'm fixing a bug inside the `ConstAccel` parametrizer.

In a nutshell, I hope you agree that this snippet of code should never fail, assuming `p` is a valid `ConstAccel` parametrizer object.
```cpp
auto bound = p.pathInterval();
auto qs = p.eval_single(bound[1]); <-- this should never throw an error, but it sometimes does on the current tip of develop
```

Reply:
I've added a comprehensive description of the bug above, and my reasoning for the content of this PR. I hope this helps clear up any confusion :D

Let me know if you have any questions.

Reply:
Thanks for the edits, it's clear for me now. 

However I think we can allow the user to provide slightly out-of-bound `times`, so that small numerical errors are acceptable (1e-10 level). In particular, 
```cpp
auto interval = pathInterval_impl();
if ((times.array() < (interval[0] * (1 - EPS_REL) - EPS_ABS)).any())
  throw std::runtime_error("a meaningful message");
// similar to the upper bound
```

The clamping can be kept as is, as we know "mathematically" that `ss` is now valid if no exceptions were raised.

Reply:
Hi @stevegolton, gentle ping. Do you still need this PR to be merged?

Reply:
Hey @hungpham2511, apologies, I got a little stuck on the updates you suggested above.

I'm happy with the PR as-is because it fixes the bug. I understand what EPS_REL and EPS_ABS is for, but I don't understand what problem it is trying to solve, so I don't know how to effectively test it or what values to choose.

The only example I can think of, a naive usage of `ConstAccel` might be something like below. Is this what you want to protect against?

```cpp
int SEGCOUNT = 10;
auto pathlen = p.pathInterval()[1] - p.pathInterval()[0];
auto seglen = pathlen / SEGCOUNT;
for (int i = 0; i <= SEGCOUNT; i++) {
  auto t = p.pathInterval()[0] + i * seglen; // <-- t might be > p.pathInterval()[1] when i == SEGCOUNT
  p.eval_single(t);
}
```

Please do what you think best with this PR, feel free to close it or add the REL/ABS changes yourself if you think it's worth keeping the bugfix. Might I suggest you merge this PR, and we address the REL/ABS issue in a separate PR?

Reply:
Hi @stevegolton, thanks for the work anyway. I think I understand your blocking point so I will work on this in the next few days to get it to merge to develop. Thanks again for the work.

Reply:
Thanks @hungpham2511 for all your help. I'm happy with how this turned out.